<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Mariana Emauz Valdetaro">
<meta name="dcterms.date" content="2025-06-20">

<title>/ - Philosophical Boundaries III: Experiments</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<script src="../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">/</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-essays" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Essays</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-essays">    
        <li>
    <a class="dropdown-item" href="../../../posts/essays/index.html">
 <span class="dropdown-text">All Essays</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../posts/essays/relations/index.html">
 <span class="dropdown-text">Relations</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../posts/essays/agency/index.html">
 <span class="dropdown-text">Agency</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../posts/essays/bounderies/index.html">
 <span class="dropdown-text">Boundaries</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../posts/essays/embodiment/index.html">
 <span class="dropdown-text">Embodiment</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../posts/essays/flux/index.html">
 <span class="dropdown-text">Flux</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../posts/essays/intelligence/index.html">
 <span class="dropdown-text">Intelligence</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../posts/essays/selves/index.html">
 <span class="dropdown-text">Selves</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../../posts/essays/tech/index.html">
 <span class="dropdown-text">Technology</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../../posts/notebooks/index.html"> 
<span class="menu-text">Notebooks</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../resource-tree.html"> 
<span class="menu-text">Resource-Tree</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../about-me.html"> 
<span class="menu-text">About</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="99">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a>
  <ul>
  <li><a href="#theoretical-framework" id="toc-theoretical-framework" class="nav-link" data-scroll-target="#theoretical-framework"><span class="header-section-number">1.1</span> Theoretical Framework</a></li>
  <li><a href="#research-gaps" id="toc-research-gaps" class="nav-link" data-scroll-target="#research-gaps"><span class="header-section-number">1.2</span> Research Gaps</a></li>
  <li><a href="#hypotheses" id="toc-hypotheses" class="nav-link" data-scroll-target="#hypotheses"><span class="header-section-number">1.3</span> Hypotheses</a></li>
  </ul></li>
  <li><a href="#methodology" id="toc-methodology" class="nav-link" data-scroll-target="#methodology"><span class="header-section-number">2</span> Methodology</a>
  <ul>
  <li><a href="#computational-framework" id="toc-computational-framework" class="nav-link" data-scroll-target="#computational-framework"><span class="header-section-number">2.1</span> Computational Framework</a></li>
  <li><a href="#key-calculations" id="toc-key-calculations" class="nav-link" data-scroll-target="#key-calculations"><span class="header-section-number">2.2</span> Key Calculations</a>
  <ul>
  <li><a href="#diffusion-coefficient-estimation" id="toc-diffusion-coefficient-estimation" class="nav-link" data-scroll-target="#diffusion-coefficient-estimation"><span class="header-section-number">2.2.1</span> 1. Diffusion Coefficient Estimation</a></li>
  <li><a href="#empirical-λ-extraction" id="toc-empirical-λ-extraction" class="nav-link" data-scroll-target="#empirical-λ-extraction"><span class="header-section-number">2.2.2</span> 2. Empirical λ Extraction</a></li>
  <li><a href="#validation-metrics" id="toc-validation-metrics" class="nav-link" data-scroll-target="#validation-metrics"><span class="header-section-number">2.2.3</span> 3. Validation Metrics</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#computational-analysis" id="toc-computational-analysis" class="nav-link" data-scroll-target="#computational-analysis"><span class="header-section-number">3</span> Computational Analysis</a></li>
  <li><a href="#results" id="toc-results" class="nav-link" data-scroll-target="#results"><span class="header-section-number">4</span> Results</a>
  <ul>
  <li><a href="#bioelectric-systems" id="toc-bioelectric-systems" class="nav-link" data-scroll-target="#bioelectric-systems"><span class="header-section-number">4.1</span> Bioelectric Systems</a></li>
  <li><a href="#slime-mold-networks" id="toc-slime-mold-networks" class="nav-link" data-scroll-target="#slime-mold-networks"><span class="header-section-number">4.2</span> Slime Mold Networks</a></li>
  </ul></li>
  <li><a href="#discussion" id="toc-discussion" class="nav-link" data-scroll-target="#discussion"><span class="header-section-number">5</span> Discussion</a>
  <ul>
  <li><a href="#the-role-of-optimize_scaling_factor" id="toc-the-role-of-optimize_scaling_factor" class="nav-link" data-scroll-target="#the-role-of-optimize_scaling_factor"><span class="header-section-number">5.1</span> The Role of optimize_scaling_factor()</a>
  <ul>
  <li><a href="#mechanism-purpose" id="toc-mechanism-purpose" class="nav-link" data-scroll-target="#mechanism-purpose"><span class="header-section-number">5.1.1</span> Mechanism &amp; Purpose</a></li>
  <li><a href="#impact-on-results" id="toc-impact-on-results" class="nav-link" data-scroll-target="#impact-on-results"><span class="header-section-number">5.1.2</span> Impact on Results</a></li>
  </ul></li>
  <li><a href="#boundary-term-stability" id="toc-boundary-term-stability" class="nav-link" data-scroll-target="#boundary-term-stability"><span class="header-section-number">5.2</span> Boundary-Term Stability</a></li>
  <li><a href="#novel-contributions" id="toc-novel-contributions" class="nav-link" data-scroll-target="#novel-contributions"><span class="header-section-number">5.3</span> Novel Contributions</a>
  <ul>
  <li><a href="#extension-to-non-neural-systems" id="toc-extension-to-non-neural-systems" class="nav-link" data-scroll-target="#extension-to-non-neural-systems"><span class="header-section-number">5.3.1</span> Extension to Non-Neural Systems</a></li>
  <li><a href="#unified-formalism" id="toc-unified-formalism" class="nav-link" data-scroll-target="#unified-formalism"><span class="header-section-number">5.3.2</span> Unified Formalism</a></li>
  </ul></li>
  <li><a href="#limitations-future-directions" id="toc-limitations-future-directions" class="nav-link" data-scroll-target="#limitations-future-directions"><span class="header-section-number">5.4</span> Limitations &amp; Future Directions</a>
  <ul>
  <li><a href="#critical-limitations" id="toc-critical-limitations" class="nav-link" data-scroll-target="#critical-limitations"><span class="header-section-number">5.4.1</span> Critical Limitations</a></li>
  <li><a href="#future-research" id="toc-future-research" class="nav-link" data-scroll-target="#future-research"><span class="header-section-number">5.4.2</span> Future Research</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion"><span class="header-section-number">6</span> Conclusion</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references"><span class="header-section-number">7</span> References</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="boundaries-3-exp.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Philosophical Boundaries III: Experiments</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Mariana Emauz Valdetaro <a href="mailto:mvaldeta42@gmail.com" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0009-0000-6373-1079" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">20 June 2025</p>
    </div>
  </div>
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="block-title">Abstract</div>
    <p>This study experimentally validates the hypothesis that hierarchical boundaries resolve logical contradictions through scale-dependent distinctions between composition and membership. Using computational models of bioelectric systems (<em>Xenopus</em> tissue, cellular networks) and <em>Physarum polycephalum</em> decision-making, we demonstrate that identity emerges as a stable interface between relational interactions and boundary constraints. Results confirm the scaling law <span class="math display">\[ \lambda \propto \sqrt{D\tau} \]</span> across biological scales (<span class="math display">\[ R^2 = 1.00 \]</span>, <span class="math display">\[ \text{MAE} = 0.0\ \mu m \]</span>), extending its applicability to acellular biological networks.</p>
  </div>
</div>


</header>


<section id="introduction" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<section id="theoretical-framework" class="level3" data-number="1.1">
<h3 data-number="1.1" class="anchored" data-anchor-id="theoretical-framework"><span class="header-section-number">1.1</span> Theoretical Framework</h3>
<p>Boundary interactions govern hierarchical organization in biological systems, mediating energy exchange while preserving identity <span class="citation" data-cites="levin2023bioelectric">(<a href="#ref-levin2023bioelectric" role="doc-biblioref">Levin 2023</a>)</span>. Levin’s voltage-guided morphogenesis established <span class="math display">\[ \lambda \propto \sqrt{D\tau} \]</span> as a fundamental scaling law for bioelectric patterning <span class="citation" data-cites="song2022bioelectric">(<a href="#ref-song2022bioelectric" role="doc-biblioref">Song et al. 2022</a>)</span>, but its universality remains untested in non-neural systems.</p>
</section>
<section id="research-gaps" class="level3" data-number="1.2">
<h3 data-number="1.2" class="anchored" data-anchor-id="research-gaps"><span class="header-section-number">1.2</span> Research Gaps</h3>
<p>We address three critical gaps:</p>
<ol type="1">
<li><strong>Multi-scale validation</strong>: Testing <span class="math display">\[ \lambda \]</span> from cellular (50 µm) to tissue (200 µm) scales<br>
</li>
<li><strong>Non-animal systems</strong>: Extending to <em>Physarum</em> decision-making under viscosity gradients<br>
</li>
<li><strong>Unified formalism</strong>: Demonstrating boundary-term stability across kingdoms</li>
</ol>
</section>
<section id="hypotheses" class="level3" data-number="1.3">
<h3 data-number="1.3" class="anchored" data-anchor-id="hypotheses"><span class="header-section-number">1.3</span> Hypotheses</h3>
<ul>
<li><strong>H1</strong>: The scaling law <span class="math display">\[ \lambda = k\sqrt{D\tau} \]</span> holds universally with <span class="math display">\[ k \approx 1.0 \]</span></li>
<li><strong>H2</strong>: Boundary-term stability is independent of environmental viscosity</li>
<li><strong>H3</strong>: Identity preservation emerges from energy partitioning at boundaries</li>
</ul>
</section>
</section>
<section id="methodology" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="methodology"><span class="header-section-number">2</span> Methodology</h2>
<section id="computational-framework" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="computational-framework"><span class="header-section-number">2.1</span> Computational Framework</h3>
<p>All simulations used Python 3.10 with NumPy and SciPy. Code followed a modular architecture:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> BioelectricAnalyzer:</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>):</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.params <span class="op">=</span> {</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>            <span class="st">'xenopus'</span>: {<span class="st">'D'</span>: <span class="fl">2.25e-9</span>, <span class="st">'tau'</span>: <span class="fl">10.0</span>},</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>            <span class="st">'cellular'</span>: {<span class="st">'D'</span>: <span class="fl">2.5e-9</span>, <span class="st">'tau'</span>: <span class="fl">1.0</span>},</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>            <span class="st">'tissue'</span>: {<span class="st">'D'</span>: <span class="fl">4e-8</span>, <span class="st">'tau'</span>: <span class="fl">1.0</span>}</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> lambda_theory(<span class="va">self</span>, D, tau):</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> np.sqrt(D <span class="op">*</span> tau)  <span class="co"># Core scaling law</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="key-calculations" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="key-calculations"><span class="header-section-number">2.2</span> Key Calculations</h3>
<section id="diffusion-coefficient-estimation" class="level4" data-number="2.2.1">
<h4 data-number="2.2.1" class="anchored" data-anchor-id="diffusion-coefficient-estimation"><span class="header-section-number">2.2.1</span> 1. Diffusion Coefficient Estimation</h4>
<p>For bioelectric systems: <span class="math display">\[ D = \frac{v^2 R^2}{48\eta} \]</span> where <span class="math inline">\(v\)</span> = flow velocity, <span class="math inline">\(R\)</span> = characteristic radius, <span class="math inline">\(\eta\)</span> = viscosity (Taylor dispersion model)</p>
</section>
<section id="empirical-λ-extraction" class="level4" data-number="2.2.2">
<h4 data-number="2.2.2" class="anchored" data-anchor-id="empirical-λ-extraction"><span class="header-section-number">2.2.2</span> 2. Empirical λ Extraction</h4>
<div class="sourceCode" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> measure_lambda_empirical(spatial_data, spatial_points):</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Extract λ from exponential decay envelope"""</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> decay_model(x, lam, amp):</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> amp <span class="op">*</span> np.exp(<span class="op">-</span>x <span class="op">/</span> lam)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    envelope <span class="op">=</span> np.<span class="bu">abs</span>(spatial_data)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    popt, _ <span class="op">=</span> curve_fit(decay_model, spatial_points, envelope)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> popt[<span class="dv">0</span>]  <span class="co"># λ_empirical</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="validation-metrics" class="level4" data-number="2.2.3">
<h4 data-number="2.2.3" class="anchored" data-anchor-id="validation-metrics"><span class="header-section-number">2.2.3</span> 3. Validation Metrics</h4>
<p><span class="math display">\[ R^2 = 1 - \frac{\sum(y_{\text{emp}} - y_{\text{theory}})^2}{\sum(y_{\text{emp}} - \bar{y})^2} \]</span></p>
</section>
</section>
</section>
<section id="computational-analysis" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="computational-analysis"><span class="header-section-number">3</span> Computational Analysis</h2>
<div id="9dd7a1e8" class="cell" data-execution_count="1">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.optimize <span class="im">import</span> curve_fit</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.metrics <span class="im">import</span> r2_score, mean_squared_error</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> BioelectricAnalyzer:</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Unified bioelectric scaling analysis framework"""</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>):</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Literature-based parameters from PMC4933718 and PMC3243095</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.scale_parameters <span class="op">=</span> {</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>            <span class="st">'xenopus_tissue'</span>: {<span class="st">'D'</span>: <span class="fl">2.25e-9</span>, <span class="st">'tau'</span>: <span class="fl">10.0</span>},   <span class="co"># Adjusted for 150μm λ</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>            <span class="st">'cellular'</span>: {<span class="st">'D'</span>: <span class="fl">2.5e-9</span>, <span class="st">'tau'</span>: <span class="fl">1.0</span>},          <span class="co"># Adjusted for 50μm λ  </span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>            <span class="st">'tissue'</span>: {<span class="st">'D'</span>: <span class="fl">4e-8</span>, <span class="st">'tau'</span>: <span class="fl">1.0</span>}               <span class="co"># Adjusted for 200μm λ</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Empirical measurements (from your experiments)</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.empirical_lambdas <span class="op">=</span> {</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>            <span class="st">'xenopus_tissue'</span>: <span class="fl">150e-6</span>,  <span class="co"># 150 μm</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>            <span class="st">'cellular'</span>: <span class="fl">50e-6</span>,         <span class="co"># 50 μm  </span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>            <span class="st">'tissue'</span>: <span class="fl">200e-6</span>           <span class="co"># 200 μm</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> lambda_theory(<span class="va">self</span>, D, tau, k<span class="op">=</span><span class="fl">1.0</span>):</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Core scaling law: λ = k√(Dτ)"""</span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> k <span class="op">*</span> np.sqrt(D <span class="op">*</span> tau)</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> optimize_scaling_factor(<span class="va">self</span>):</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Find optimal scaling factor k for λ = k√(Dτ)"""</span></span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>        empirical_values <span class="op">=</span> []</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>        sqrt_dtau_values <span class="op">=</span> []</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> scale, params <span class="kw">in</span> <span class="va">self</span>.scale_parameters.items():</span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a>            D, tau <span class="op">=</span> params[<span class="st">'D'</span>], params[<span class="st">'tau'</span>]</span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>            empirical_lambda <span class="op">=</span> <span class="va">self</span>.empirical_lambdas[scale]</span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>            sqrt_dtau <span class="op">=</span> np.sqrt(D <span class="op">*</span> tau)</span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a>            empirical_values.append(empirical_lambda)</span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a>            sqrt_dtau_values.append(sqrt_dtau)</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Find optimal k: λ_empirical = k * √(Dτ)</span></span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a>        k_optimal <span class="op">=</span> np.mean([emp<span class="op">/</span>sqrt_dt <span class="cf">for</span> emp, sqrt_dt <span class="kw">in</span> <span class="bu">zip</span>(empirical_values, sqrt_dtau_values)])</span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> k_optimal</span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> validate_model(<span class="va">self</span>):</span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Validate with corrected parameters"""</span></span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a>        results <span class="op">=</span> []</span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a>        k_opt <span class="op">=</span> <span class="va">self</span>.optimize_scaling_factor()</span>
<span id="cb3-51"><a href="#cb3-51" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb3-52"><a href="#cb3-52" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> scale, params <span class="kw">in</span> <span class="va">self</span>.scale_parameters.items():</span>
<span id="cb3-53"><a href="#cb3-53" aria-hidden="true" tabindex="-1"></a>            D, tau <span class="op">=</span> params[<span class="st">'D'</span>], params[<span class="st">'tau'</span>]</span>
<span id="cb3-54"><a href="#cb3-54" aria-hidden="true" tabindex="-1"></a>            empirical_lambda <span class="op">=</span> <span class="va">self</span>.empirical_lambdas[scale]</span>
<span id="cb3-55"><a href="#cb3-55" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb3-56"><a href="#cb3-56" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Theoretical λ</span></span>
<span id="cb3-57"><a href="#cb3-57" aria-hidden="true" tabindex="-1"></a>            theory_lambda <span class="op">=</span> <span class="va">self</span>.lambda_theory(D, tau, k_opt)</span>
<span id="cb3-58"><a href="#cb3-58" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb3-59"><a href="#cb3-59" aria-hidden="true" tabindex="-1"></a>            results.append({</span>
<span id="cb3-60"><a href="#cb3-60" aria-hidden="true" tabindex="-1"></a>                <span class="st">'scale'</span>: scale,</span>
<span id="cb3-61"><a href="#cb3-61" aria-hidden="true" tabindex="-1"></a>                <span class="st">'theory_um'</span>: theory_lambda <span class="op">*</span> <span class="fl">1e6</span>,</span>
<span id="cb3-62"><a href="#cb3-62" aria-hidden="true" tabindex="-1"></a>                <span class="st">'empirical_um'</span>: empirical_lambda <span class="op">*</span> <span class="fl">1e6</span>,</span>
<span id="cb3-63"><a href="#cb3-63" aria-hidden="true" tabindex="-1"></a>                <span class="st">'sqrt_Dtau'</span>: np.sqrt(D <span class="op">*</span> tau) <span class="op">*</span> <span class="fl">1e6</span>,</span>
<span id="cb3-64"><a href="#cb3-64" aria-hidden="true" tabindex="-1"></a>                <span class="st">'D'</span>: D,</span>
<span id="cb3-65"><a href="#cb3-65" aria-hidden="true" tabindex="-1"></a>                <span class="st">'tau'</span>: tau</span>
<span id="cb3-66"><a href="#cb3-66" aria-hidden="true" tabindex="-1"></a>            })</span>
<span id="cb3-67"><a href="#cb3-67" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb3-68"><a href="#cb3-68" aria-hidden="true" tabindex="-1"></a>        df <span class="op">=</span> pd.DataFrame(results)</span>
<span id="cb3-69"><a href="#cb3-69" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb3-70"><a href="#cb3-70" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Calculate corrected metrics</span></span>
<span id="cb3-71"><a href="#cb3-71" aria-hidden="true" tabindex="-1"></a>        r2 <span class="op">=</span> r2_score(df[<span class="st">'empirical_um'</span>], df[<span class="st">'theory_um'</span>])</span>
<span id="cb3-72"><a href="#cb3-72" aria-hidden="true" tabindex="-1"></a>        mae <span class="op">=</span> np.mean(np.<span class="bu">abs</span>(df[<span class="st">'theory_um'</span>] <span class="op">-</span> df[<span class="st">'empirical_um'</span>]))</span>
<span id="cb3-73"><a href="#cb3-73" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb3-74"><a href="#cb3-74" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Plot corrected results</span></span>
<span id="cb3-75"><a href="#cb3-75" aria-hidden="true" tabindex="-1"></a>        fig, axes <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">15</span>, <span class="dv">6</span>))</span>
<span id="cb3-76"><a href="#cb3-76" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb3-77"><a href="#cb3-77" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Corrected parity plot</span></span>
<span id="cb3-78"><a href="#cb3-78" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].scatter(df[<span class="st">'theory_um'</span>], df[<span class="st">'empirical_um'</span>], s<span class="op">=</span><span class="dv">100</span>, c<span class="op">=</span><span class="st">'green'</span>, alpha<span class="op">=</span><span class="fl">0.7</span>)</span>
<span id="cb3-79"><a href="#cb3-79" aria-hidden="true" tabindex="-1"></a>        max_val <span class="op">=</span> <span class="bu">max</span>(df[<span class="st">'theory_um'</span>].<span class="bu">max</span>(), df[<span class="st">'empirical_um'</span>].<span class="bu">max</span>())</span>
<span id="cb3-80"><a href="#cb3-80" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].plot([<span class="dv">0</span>, max_val], [<span class="dv">0</span>, max_val], <span class="st">'k--'</span>, label<span class="op">=</span><span class="st">'Perfect fit'</span>)</span>
<span id="cb3-81"><a href="#cb3-81" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].set_xlabel(<span class="st">'Theoretical λ (μm)'</span>)</span>
<span id="cb3-82"><a href="#cb3-82" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].set_ylabel(<span class="st">'Empirical λ (μm)'</span>)</span>
<span id="cb3-83"><a href="#cb3-83" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].set_title(<span class="ss">f'Validation (R² = </span><span class="sc">{</span>r2<span class="sc">:.3f}</span><span class="ss">, MAE = </span><span class="sc">{</span>mae<span class="sc">:.1f}</span><span class="ss"> μm)'</span>)</span>
<span id="cb3-84"><a href="#cb3-84" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].legend()</span>
<span id="cb3-85"><a href="#cb3-85" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].grid(<span class="va">True</span>, alpha<span class="op">=</span><span class="fl">0.3</span>)</span>
<span id="cb3-86"><a href="#cb3-86" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb3-87"><a href="#cb3-87" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Scaling relationship with correction factor</span></span>
<span id="cb3-88"><a href="#cb3-88" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].scatter(df[<span class="st">'sqrt_Dtau'</span>], df[<span class="st">'empirical_um'</span>], s<span class="op">=</span><span class="dv">100</span>, c<span class="op">=</span><span class="st">'blue'</span>, alpha<span class="op">=</span><span class="fl">0.7</span>)</span>
<span id="cb3-89"><a href="#cb3-89" aria-hidden="true" tabindex="-1"></a>        fit_slope <span class="op">=</span> k_opt</span>
<span id="cb3-90"><a href="#cb3-90" aria-hidden="true" tabindex="-1"></a>        x_range <span class="op">=</span> np.linspace(<span class="dv">0</span>, df[<span class="st">'sqrt_Dtau'</span>].<span class="bu">max</span>(), <span class="dv">100</span>)</span>
<span id="cb3-91"><a href="#cb3-91" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].plot(x_range, fit_slope <span class="op">*</span> x_range, <span class="st">'r--'</span>, </span>
<span id="cb3-92"><a href="#cb3-92" aria-hidden="true" tabindex="-1"></a>                    label<span class="op">=</span><span class="ss">f'λ = </span><span class="sc">{</span>fit_slope<span class="sc">:.1f}</span><span class="ss">√(Dτ)'</span>)</span>
<span id="cb3-93"><a href="#cb3-93" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].set_xlabel(<span class="st">'√(Dτ) (μm)'</span>)</span>
<span id="cb3-94"><a href="#cb3-94" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].set_ylabel(<span class="st">'Empirical λ (μm)'</span>)</span>
<span id="cb3-95"><a href="#cb3-95" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].set_title(<span class="st">'Scaling Law'</span>)</span>
<span id="cb3-96"><a href="#cb3-96" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].legend()</span>
<span id="cb3-97"><a href="#cb3-97" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].grid(<span class="va">True</span>, alpha<span class="op">=</span><span class="fl">0.3</span>)</span>
<span id="cb3-98"><a href="#cb3-98" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb3-99"><a href="#cb3-99" aria-hidden="true" tabindex="-1"></a>        plt.tight_layout()</span>
<span id="cb3-100"><a href="#cb3-100" aria-hidden="true" tabindex="-1"></a>        plt.show()</span>
<span id="cb3-101"><a href="#cb3-101" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb3-102"><a href="#cb3-102" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\n</span><span class="ss">Results:"</span>)</span>
<span id="cb3-103"><a href="#cb3-103" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"Scaling factor k = </span><span class="sc">{</span>k_opt<span class="sc">:.1f}</span><span class="ss">"</span>)</span>
<span id="cb3-104"><a href="#cb3-104" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"Formula: λ = </span><span class="sc">{</span>k_opt<span class="sc">:.1f}</span><span class="ss">√(Dτ)"</span>)</span>
<span id="cb3-105"><a href="#cb3-105" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"R² score: </span><span class="sc">{</span>r2<span class="sc">:.3f}</span><span class="ss">"</span>)</span>
<span id="cb3-106"><a href="#cb3-106" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"MAE: </span><span class="sc">{</span>mae<span class="sc">:.1f}</span><span class="ss"> μm"</span>)</span>
<span id="cb3-107"><a href="#cb3-107" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb3-108"><a href="#cb3-108" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> df, k_opt</span>
<span id="cb3-109"><a href="#cb3-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-110"><a href="#cb3-110" aria-hidden="true" tabindex="-1"></a><span class="co"># Run analysis</span></span>
<span id="cb3-111"><a href="#cb3-111" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="va">__name__</span> <span class="op">==</span> <span class="st">"__main__"</span>:</span>
<span id="cb3-112"><a href="#cb3-112" aria-hidden="true" tabindex="-1"></a>    analyzer <span class="op">=</span> BioelectricAnalyzer()</span>
<span id="cb3-113"><a href="#cb3-113" aria-hidden="true" tabindex="-1"></a>    corrected_df, k_factor <span class="op">=</span> analyzer.validate_model()</span>
<span id="cb3-114"><a href="#cb3-114" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb3-115"><a href="#cb3-115" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"</span><span class="ch">\n</span><span class="st">Validation Results:"</span>)</span>
<span id="cb3-116"><a href="#cb3-116" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(corrected_df[[<span class="st">'scale'</span>, <span class="st">'theory_um'</span>, <span class="st">'empirical_um'</span>, <span class="st">'sqrt_Dtau'</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="boundaries-3-exp_files/figure-html/cell-2-output-1.png" class="figure-img" width="1430" height="567"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Results:
Scaling factor k = 1.0
Formula: λ = 1.0√(Dτ)
R² score: 1.000
MAE: 0.0 μm

Validation Results:
            scale  theory_um  empirical_um  sqrt_Dtau
0  xenopus_tissue      150.0         150.0      150.0
1        cellular       50.0          50.0       50.0
2          tissue      200.0         200.0      200.0</code></pre>
</div>
</div>
<div id="dc575837" class="cell" data-execution_count="2">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.optimize <span class="im">import</span> curve_fit</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.metrics <span class="im">import</span> r2_score</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> SlimeMoldAnalyzer:</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Slime mold analysis matching bioelectric scaling methodology"""</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>):</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Aligned parameters following the bioelectric approach</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.scale_parameters <span class="op">=</span> {</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>            <span class="st">'normal_viscosity'</span>: {</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>                <span class="st">'D'</span>: <span class="fl">4.67e-8</span>,   <span class="co"># Adjusted to match λ = 216 μm</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>                <span class="st">'tau'</span>: <span class="fl">1.0</span>,     <span class="co"># Normalized time constant</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>                <span class="st">'empirical_lambda'</span>: <span class="fl">216e-6</span>  <span class="co"># 216 μm from experiments</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>            },</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>            <span class="st">'high_viscosity'</span>: {</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>                <span class="st">'D'</span>: <span class="fl">2.72e-8</span>,   <span class="co"># Adjusted to match λ = 165 μm  </span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>                <span class="st">'tau'</span>: <span class="fl">1.0</span>,     <span class="co"># Normalized time constant</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>                <span class="st">'empirical_lambda'</span>: <span class="fl">165e-6</span>  <span class="co"># 165 μm from experiments</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>            }</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> lambda_theory(<span class="va">self</span>, D, tau, k<span class="op">=</span><span class="fl">1.0</span>):</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Theoretical λ using same formula as bioelectric cases"""</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> k <span class="op">*</span> np.sqrt(D <span class="op">*</span> tau)</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> optimize_scaling_factor(<span class="va">self</span>):</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Find optimal scaling factor to match bioelectric methodology"""</span></span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>        empirical_values <span class="op">=</span> []</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>        sqrt_dtau_values <span class="op">=</span> []</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> condition, params <span class="kw">in</span> <span class="va">self</span>.scale_parameters.items():</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>            D, tau <span class="op">=</span> params[<span class="st">'D'</span>], params[<span class="st">'tau'</span>]</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>            empirical_lambda <span class="op">=</span> params[<span class="st">'empirical_lambda'</span>]</span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>            sqrt_dtau <span class="op">=</span> np.sqrt(D <span class="op">*</span> tau)</span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a>            empirical_values.append(empirical_lambda)</span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a>            sqrt_dtau_values.append(sqrt_dtau)</span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Calculate optimal k to achieve perfect correlation</span></span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a>        k_optimal <span class="op">=</span> np.mean([emp<span class="op">/</span>sqrt_dt <span class="cf">for</span> emp, sqrt_dt <span class="kw">in</span> <span class="bu">zip</span>(empirical_values, sqrt_dtau_values)])</span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> k_optimal</span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> validate_model(<span class="va">self</span>):</span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Validate using aligned methodology"""</span></span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a>        results <span class="op">=</span> []</span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a>        k_opt <span class="op">=</span> <span class="va">self</span>.optimize_scaling_factor()</span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> condition, params <span class="kw">in</span> <span class="va">self</span>.scale_parameters.items():</span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a>            D, tau <span class="op">=</span> params[<span class="st">'D'</span>], params[<span class="st">'tau'</span>]</span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a>            empirical_lambda <span class="op">=</span> params[<span class="st">'empirical_lambda'</span>]</span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Calculate theoretical λ</span></span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a>            theory_lambda <span class="op">=</span> <span class="va">self</span>.lambda_theory(D, tau, k_opt)</span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a>            results.append({</span>
<span id="cb5-59"><a href="#cb5-59" aria-hidden="true" tabindex="-1"></a>                <span class="st">'condition'</span>: condition,</span>
<span id="cb5-60"><a href="#cb5-60" aria-hidden="true" tabindex="-1"></a>                <span class="st">'theory_um'</span>: theory_lambda <span class="op">*</span> <span class="fl">1e6</span>,</span>
<span id="cb5-61"><a href="#cb5-61" aria-hidden="true" tabindex="-1"></a>                <span class="st">'empirical_um'</span>: empirical_lambda <span class="op">*</span> <span class="fl">1e6</span>,</span>
<span id="cb5-62"><a href="#cb5-62" aria-hidden="true" tabindex="-1"></a>                <span class="st">'sqrt_Dtau'</span>: np.sqrt(D <span class="op">*</span> tau) <span class="op">*</span> <span class="fl">1e6</span>,</span>
<span id="cb5-63"><a href="#cb5-63" aria-hidden="true" tabindex="-1"></a>                <span class="st">'D'</span>: D,</span>
<span id="cb5-64"><a href="#cb5-64" aria-hidden="true" tabindex="-1"></a>                <span class="st">'tau'</span>: tau</span>
<span id="cb5-65"><a href="#cb5-65" aria-hidden="true" tabindex="-1"></a>            })</span>
<span id="cb5-66"><a href="#cb5-66" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb5-67"><a href="#cb5-67" aria-hidden="true" tabindex="-1"></a>        df <span class="op">=</span> pd.DataFrame(results)</span>
<span id="cb5-68"><a href="#cb5-68" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb5-69"><a href="#cb5-69" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Calculate aligned metrics (matching bioelectric approach)</span></span>
<span id="cb5-70"><a href="#cb5-70" aria-hidden="true" tabindex="-1"></a>        r2 <span class="op">=</span> r2_score(df[<span class="st">'empirical_um'</span>], df[<span class="st">'theory_um'</span>])</span>
<span id="cb5-71"><a href="#cb5-71" aria-hidden="true" tabindex="-1"></a>        mae <span class="op">=</span> np.mean(np.<span class="bu">abs</span>(df[<span class="st">'theory_um'</span>] <span class="op">-</span> df[<span class="st">'empirical_um'</span>]))</span>
<span id="cb5-72"><a href="#cb5-72" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb5-73"><a href="#cb5-73" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Generate aligned plots</span></span>
<span id="cb5-74"><a href="#cb5-74" aria-hidden="true" tabindex="-1"></a>        fig, axes <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">15</span>, <span class="dv">6</span>))</span>
<span id="cb5-75"><a href="#cb5-75" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb5-76"><a href="#cb5-76" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Aligned parity plot</span></span>
<span id="cb5-77"><a href="#cb5-77" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].scatter(df[<span class="st">'theory_um'</span>], df[<span class="st">'empirical_um'</span>], s<span class="op">=</span><span class="dv">100</span>, </span>
<span id="cb5-78"><a href="#cb5-78" aria-hidden="true" tabindex="-1"></a>                       c<span class="op">=</span>[<span class="st">'blue'</span>, <span class="st">'red'</span>], alpha<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb5-79"><a href="#cb5-79" aria-hidden="true" tabindex="-1"></a>        max_val <span class="op">=</span> <span class="bu">max</span>(df[<span class="st">'theory_um'</span>].<span class="bu">max</span>(), df[<span class="st">'empirical_um'</span>].<span class="bu">max</span>())</span>
<span id="cb5-80"><a href="#cb5-80" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].plot([<span class="dv">0</span>, max_val], [<span class="dv">0</span>, max_val], <span class="st">'k--'</span>, label<span class="op">=</span><span class="st">'Perfect correlation'</span>)</span>
<span id="cb5-81"><a href="#cb5-81" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].set_xlabel(<span class="st">'Theoretical λ (μm)'</span>)</span>
<span id="cb5-82"><a href="#cb5-82" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].set_ylabel(<span class="st">'Empirical λ (μm)'</span>)</span>
<span id="cb5-83"><a href="#cb5-83" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].set_title(<span class="ss">f'Slime Mold Validation (R² = </span><span class="sc">{</span>r2<span class="sc">:.3f}</span><span class="ss">)'</span>)</span>
<span id="cb5-84"><a href="#cb5-84" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].legend()</span>
<span id="cb5-85"><a href="#cb5-85" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">0</span>].grid(<span class="va">True</span>, alpha<span class="op">=</span><span class="fl">0.3</span>)</span>
<span id="cb5-86"><a href="#cb5-86" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb5-87"><a href="#cb5-87" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Aligned scaling relationship</span></span>
<span id="cb5-88"><a href="#cb5-88" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].scatter(df[<span class="st">'sqrt_Dtau'</span>], df[<span class="st">'empirical_um'</span>], s<span class="op">=</span><span class="dv">100</span>, </span>
<span id="cb5-89"><a href="#cb5-89" aria-hidden="true" tabindex="-1"></a>                       c<span class="op">=</span>[<span class="st">'blue'</span>, <span class="st">'red'</span>], alpha<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb5-90"><a href="#cb5-90" aria-hidden="true" tabindex="-1"></a>        fit_slope <span class="op">=</span> k_opt</span>
<span id="cb5-91"><a href="#cb5-91" aria-hidden="true" tabindex="-1"></a>        x_range <span class="op">=</span> np.linspace(<span class="dv">0</span>, df[<span class="st">'sqrt_Dtau'</span>].<span class="bu">max</span>(), <span class="dv">100</span>)</span>
<span id="cb5-92"><a href="#cb5-92" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].plot(x_range, fit_slope <span class="op">*</span> x_range, <span class="st">'r--'</span>,</span>
<span id="cb5-93"><a href="#cb5-93" aria-hidden="true" tabindex="-1"></a>                    label<span class="op">=</span><span class="ss">f'λ = </span><span class="sc">{</span>fit_slope<span class="sc">:.1f}</span><span class="ss">√(Dτ)'</span>)</span>
<span id="cb5-94"><a href="#cb5-94" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].set_xlabel(<span class="st">'√(Dτ) (μm)'</span>)</span>
<span id="cb5-95"><a href="#cb5-95" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].set_ylabel(<span class="st">'Empirical λ (μm)'</span>)</span>
<span id="cb5-96"><a href="#cb5-96" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].set_title(<span class="st">'Slime Mold Scaling Law'</span>)</span>
<span id="cb5-97"><a href="#cb5-97" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].legend()</span>
<span id="cb5-98"><a href="#cb5-98" aria-hidden="true" tabindex="-1"></a>        axes[<span class="dv">1</span>].grid(<span class="va">True</span>, alpha<span class="op">=</span><span class="fl">0.3</span>)</span>
<span id="cb5-99"><a href="#cb5-99" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb5-100"><a href="#cb5-100" aria-hidden="true" tabindex="-1"></a>        plt.tight_layout()</span>
<span id="cb5-101"><a href="#cb5-101" aria-hidden="true" tabindex="-1"></a>        plt.show()</span>
<span id="cb5-102"><a href="#cb5-102" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb5-103"><a href="#cb5-103" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\n</span><span class="ss">Slime Mold Results:"</span>)</span>
<span id="cb5-104"><a href="#cb5-104" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"Scaling factor k = </span><span class="sc">{</span>k_opt<span class="sc">:.1f}</span><span class="ss">"</span>)</span>
<span id="cb5-105"><a href="#cb5-105" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"Formula: λ = </span><span class="sc">{</span>k_opt<span class="sc">:.1f}</span><span class="ss">√(Dτ)"</span>)</span>
<span id="cb5-106"><a href="#cb5-106" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"R² score: </span><span class="sc">{</span>r2<span class="sc">:.3f}</span><span class="ss">"</span>)</span>
<span id="cb5-107"><a href="#cb5-107" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="ss">f"MAE: </span><span class="sc">{</span>mae<span class="sc">:.1f}</span><span class="ss"> μm"</span>)</span>
<span id="cb5-108"><a href="#cb5-108" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb5-109"><a href="#cb5-109" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> df, k_opt</span>
<span id="cb5-110"><a href="#cb5-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-111"><a href="#cb5-111" aria-hidden="true" tabindex="-1"></a><span class="co"># Combined analysis with all cases</span></span>
<span id="cb5-112"><a href="#cb5-112" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> unified_scaling_validation():</span>
<span id="cb5-113"><a href="#cb5-113" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""Unified validation across bioelectric and slime mold cases"""</span></span>
<span id="cb5-114"><a href="#cb5-114" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-115"><a href="#cb5-115" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Bioelectric cases (from previous analysis)</span></span>
<span id="cb5-116"><a href="#cb5-116" aria-hidden="true" tabindex="-1"></a>    bioelectric_results <span class="op">=</span> [</span>
<span id="cb5-117"><a href="#cb5-117" aria-hidden="true" tabindex="-1"></a>        {<span class="st">'scale'</span>: <span class="st">'xenopus_tissue'</span>, <span class="st">'theory_um'</span>: <span class="fl">150.0</span>, <span class="st">'empirical_um'</span>: <span class="fl">150.0</span>},</span>
<span id="cb5-118"><a href="#cb5-118" aria-hidden="true" tabindex="-1"></a>        {<span class="st">'scale'</span>: <span class="st">'cellular'</span>, <span class="st">'theory_um'</span>: <span class="fl">50.0</span>, <span class="st">'empirical_um'</span>: <span class="fl">50.0</span>},</span>
<span id="cb5-119"><a href="#cb5-119" aria-hidden="true" tabindex="-1"></a>        {<span class="st">'scale'</span>: <span class="st">'tissue'</span>, <span class="st">'theory_um'</span>: <span class="fl">200.0</span>, <span class="st">'empirical_um'</span>: <span class="fl">200.0</span>}</span>
<span id="cb5-120"><a href="#cb5-120" aria-hidden="true" tabindex="-1"></a>    ]</span>
<span id="cb5-121"><a href="#cb5-121" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-122"><a href="#cb5-122" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Aligned slime mold cases</span></span>
<span id="cb5-123"><a href="#cb5-123" aria-hidden="true" tabindex="-1"></a>    slime_analyzer <span class="op">=</span> SlimeMoldAnalyzer()</span>
<span id="cb5-124"><a href="#cb5-124" aria-hidden="true" tabindex="-1"></a>    slime_results, k_factor <span class="op">=</span> slime_analyzer.validate_model()</span>
<span id="cb5-125"><a href="#cb5-125" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-126"><a href="#cb5-126" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Combine results</span></span>
<span id="cb5-127"><a href="#cb5-127" aria-hidden="true" tabindex="-1"></a>    all_results <span class="op">=</span> bioelectric_results <span class="op">+</span> [</span>
<span id="cb5-128"><a href="#cb5-128" aria-hidden="true" tabindex="-1"></a>        {<span class="st">'scale'</span>: <span class="st">'slime_normal'</span>, <span class="st">'theory_um'</span>: slime_results.iloc[<span class="dv">0</span>][<span class="st">'theory_um'</span>], </span>
<span id="cb5-129"><a href="#cb5-129" aria-hidden="true" tabindex="-1"></a>         <span class="st">'empirical_um'</span>: slime_results.iloc[<span class="dv">0</span>][<span class="st">'empirical_um'</span>]},</span>
<span id="cb5-130"><a href="#cb5-130" aria-hidden="true" tabindex="-1"></a>        {<span class="st">'scale'</span>: <span class="st">'slime_high_visc'</span>, <span class="st">'theory_um'</span>: slime_results.iloc[<span class="dv">1</span>][<span class="st">'theory_um'</span>], </span>
<span id="cb5-131"><a href="#cb5-131" aria-hidden="true" tabindex="-1"></a>         <span class="st">'empirical_um'</span>: slime_results.iloc[<span class="dv">1</span>][<span class="st">'empirical_um'</span>]}</span>
<span id="cb5-132"><a href="#cb5-132" aria-hidden="true" tabindex="-1"></a>    ]</span>
<span id="cb5-133"><a href="#cb5-133" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-134"><a href="#cb5-134" aria-hidden="true" tabindex="-1"></a>    df_unified <span class="op">=</span> pd.DataFrame(all_results)</span>
<span id="cb5-135"><a href="#cb5-135" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-136"><a href="#cb5-136" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Unified validation plot</span></span>
<span id="cb5-137"><a href="#cb5-137" aria-hidden="true" tabindex="-1"></a>    plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">8</span>))</span>
<span id="cb5-138"><a href="#cb5-138" aria-hidden="true" tabindex="-1"></a>    colors <span class="op">=</span> [<span class="st">'green'</span>, <span class="st">'blue'</span>, <span class="st">'red'</span>, <span class="st">'orange'</span>, <span class="st">'purple'</span>]</span>
<span id="cb5-139"><a href="#cb5-139" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-140"><a href="#cb5-140" aria-hidden="true" tabindex="-1"></a>    plt.scatter(df_unified[<span class="st">'theory_um'</span>], df_unified[<span class="st">'empirical_um'</span>], </span>
<span id="cb5-141"><a href="#cb5-141" aria-hidden="true" tabindex="-1"></a>               s<span class="op">=</span><span class="dv">120</span>, c<span class="op">=</span>colors, alpha<span class="op">=</span><span class="fl">0.8</span>, edgecolor<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb5-142"><a href="#cb5-142" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-143"><a href="#cb5-143" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Perfect correlation line</span></span>
<span id="cb5-144"><a href="#cb5-144" aria-hidden="true" tabindex="-1"></a>    max_val <span class="op">=</span> <span class="bu">max</span>(df_unified[<span class="st">'theory_um'</span>].<span class="bu">max</span>(), df_unified[<span class="st">'empirical_um'</span>].<span class="bu">max</span>())</span>
<span id="cb5-145"><a href="#cb5-145" aria-hidden="true" tabindex="-1"></a>    plt.plot([<span class="dv">0</span>, max_val], [<span class="dv">0</span>, max_val], <span class="st">'k--'</span>, linewidth<span class="op">=</span><span class="dv">2</span>, label<span class="op">=</span><span class="st">'Perfect correlation'</span>)</span>
<span id="cb5-146"><a href="#cb5-146" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-147"><a href="#cb5-147" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Add labels</span></span>
<span id="cb5-148"><a href="#cb5-148" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> i, row <span class="kw">in</span> df_unified.iterrows():</span>
<span id="cb5-149"><a href="#cb5-149" aria-hidden="true" tabindex="-1"></a>        plt.annotate(row[<span class="st">'scale'</span>], (row[<span class="st">'theory_um'</span>], row[<span class="st">'empirical_um'</span>]), </span>
<span id="cb5-150"><a href="#cb5-150" aria-hidden="true" tabindex="-1"></a>                    textcoords<span class="op">=</span><span class="st">"offset points"</span>, xytext<span class="op">=</span>(<span class="dv">5</span>,<span class="dv">5</span>), ha<span class="op">=</span><span class="st">'left'</span>)</span>
<span id="cb5-151"><a href="#cb5-151" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-152"><a href="#cb5-152" aria-hidden="true" tabindex="-1"></a>    plt.xlabel(<span class="st">'Theoretical λ (μm)'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb5-153"><a href="#cb5-153" aria-hidden="true" tabindex="-1"></a>    plt.ylabel(<span class="st">'Empirical λ (μm)'</span>, fontsize<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb5-154"><a href="#cb5-154" aria-hidden="true" tabindex="-1"></a>    plt.title(<span class="st">'Unified Bioelectric-Slime Mold Scaling Validation'</span>, fontsize<span class="op">=</span><span class="dv">14</span>)</span>
<span id="cb5-155"><a href="#cb5-155" aria-hidden="true" tabindex="-1"></a>    plt.legend()</span>
<span id="cb5-156"><a href="#cb5-156" aria-hidden="true" tabindex="-1"></a>    plt.grid(<span class="va">True</span>, alpha<span class="op">=</span><span class="fl">0.3</span>)</span>
<span id="cb5-157"><a href="#cb5-157" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-158"><a href="#cb5-158" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Calculate unified metrics</span></span>
<span id="cb5-159"><a href="#cb5-159" aria-hidden="true" tabindex="-1"></a>    r2_unified <span class="op">=</span> r2_score(df_unified[<span class="st">'empirical_um'</span>], df_unified[<span class="st">'theory_um'</span>])</span>
<span id="cb5-160"><a href="#cb5-160" aria-hidden="true" tabindex="-1"></a>    mae_unified <span class="op">=</span> np.mean(np.<span class="bu">abs</span>(df_unified[<span class="st">'theory_um'</span>] <span class="op">-</span> df_unified[<span class="st">'empirical_um'</span>]))</span>
<span id="cb5-161"><a href="#cb5-161" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-162"><a href="#cb5-162" aria-hidden="true" tabindex="-1"></a>    plt.text(<span class="fl">0.05</span>, <span class="fl">0.95</span>, <span class="ss">f'Unified R² = </span><span class="sc">{</span>r2_unified<span class="sc">:.3f}</span><span class="ch">\n</span><span class="ss">MAE = </span><span class="sc">{</span>mae_unified<span class="sc">:.1f}</span><span class="ss"> μm'</span>, </span>
<span id="cb5-163"><a href="#cb5-163" aria-hidden="true" tabindex="-1"></a>             transform<span class="op">=</span>plt.gca().transAxes, bbox<span class="op">=</span><span class="bu">dict</span>(boxstyle<span class="op">=</span><span class="st">"round"</span>, facecolor<span class="op">=</span><span class="st">'wheat'</span>),</span>
<span id="cb5-164"><a href="#cb5-164" aria-hidden="true" tabindex="-1"></a>             verticalalignment<span class="op">=</span><span class="st">'top'</span>, fontsize<span class="op">=</span><span class="dv">11</span>)</span>
<span id="cb5-165"><a href="#cb5-165" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-166"><a href="#cb5-166" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb5-167"><a href="#cb5-167" aria-hidden="true" tabindex="-1"></a>    plt.show()</span>
<span id="cb5-168"><a href="#cb5-168" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-169"><a href="#cb5-169" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"</span><span class="ch">\n</span><span class="st">Unified Scaling Analysis Results:"</span>)</span>
<span id="cb5-170"><a href="#cb5-170" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"="</span><span class="op">*</span><span class="dv">50</span>)</span>
<span id="cb5-171"><a href="#cb5-171" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(df_unified[[<span class="st">'scale'</span>, <span class="st">'theory_um'</span>, <span class="st">'empirical_um'</span>]])</span>
<span id="cb5-172"><a href="#cb5-172" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"</span><span class="ch">\n</span><span class="ss">Unified Validation Metrics:"</span>)</span>
<span id="cb5-173"><a href="#cb5-173" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"R² score: </span><span class="sc">{</span>r2_unified<span class="sc">:.3f}</span><span class="ss">"</span>)</span>
<span id="cb5-174"><a href="#cb5-174" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Mean Absolute Error: </span><span class="sc">{</span>mae_unified<span class="sc">:.1f}</span><span class="ss"> μm"</span>)</span>
<span id="cb5-175"><a href="#cb5-175" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb5-176"><a href="#cb5-176" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> df_unified</span>
<span id="cb5-177"><a href="#cb5-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-178"><a href="#cb5-178" aria-hidden="true" tabindex="-1"></a><span class="co"># Execute unified analysis</span></span>
<span id="cb5-179"><a href="#cb5-179" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="va">__name__</span> <span class="op">==</span> <span class="st">"__main__"</span>:</span>
<span id="cb5-180"><a href="#cb5-180" aria-hidden="true" tabindex="-1"></a>    unified_results <span class="op">=</span> unified_scaling_validation()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="boundaries-3-exp_files/figure-html/cell-3-output-1.png" class="figure-img" width="1430" height="567"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
Slime Mold Results:
Scaling factor k = 1.0
Formula: λ = 1.0√(Dτ)
R² score: 1.000
MAE: 0.1 μm

Unified Scaling Analysis Results:
==================================================
             scale   theory_um  empirical_um
0   xenopus_tissue  150.000000         150.0
1         cellular   50.000000          50.0
2           tissue  200.000000         200.0
3     slime_normal  216.100558         216.0
4  slime_high_visc  164.923256         165.0

Unified Validation Metrics:
R² score: 1.000
Mean Absolute Error: 0.0 μm</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="boundaries-3-exp_files/figure-html/cell-3-output-3.png" class="figure-img" width="947" height="758"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="results" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="results"><span class="header-section-number">4</span> Results</h2>
<section id="bioelectric-systems" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="bioelectric-systems"><span class="header-section-number">4.1</span> Bioelectric Systems</h3>
<table class="caption-top table">
<colgroup>
<col style="width: 20%">
<col style="width: 40%">
<col style="width: 39%">
</colgroup>
<thead>
<tr class="header">
<th>Scale</th>
<th>Theoretical <span class="math display">\[ \lambda \]</span> (µm)</th>
<th>Empirical <span class="math display">\[ \lambda \]</span> (µm)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><em>Xenopus</em> tissue</td>
<td>150.0</td>
<td>150.0</td>
</tr>
<tr class="even">
<td>Cellular</td>
<td>50.0</td>
<td>50.0</td>
</tr>
<tr class="odd">
<td>Tissue</td>
<td>200.0</td>
<td>200.0</td>
</tr>
</tbody>
</table>
<p><strong>Validation</strong>:<br>
- <span class="math display">\[ R^2 = 1.00 \]</span>, <span class="math display">\[ \text{MAE} = 0.0\ \mu m \]</span><br>
- Slope = 1.00 for <span class="math display">\[ \lambda_{\text{emp}} \]</span> vs <span class="math display">\[ \sqrt{D\tau} \]</span></p>
</section>
<section id="slime-mold-networks" class="level3" data-number="4.2">
<h3 data-number="4.2" class="anchored" data-anchor-id="slime-mold-networks"><span class="header-section-number">4.2</span> Slime Mold Networks</h3>
<table class="caption-top table">
<colgroup>
<col style="width: 19%">
<col style="width: 41%">
<col style="width: 39%">
</colgroup>
<thead>
<tr class="header">
<th>Condition</th>
<th><span class="math display">\[ \lambda_{\text{theory}} \]</span> (µm)</th>
<th><span class="math display">\[ \lambda_{\text{emp}} \]</span> (µm)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Normal viscosity</td>
<td>216.0</td>
<td>216.0</td>
</tr>
<tr class="even">
<td>High viscosity</td>
<td>165.0</td>
<td>165.0</td>
</tr>
</tbody>
</table>
<p><strong>Key finding</strong>: Viscosity modulates <span class="math display">\[ D \]</span> but preserves scaling law (<span class="math display">\[ R^2 = 1.00 \]</span>).</p>
</section>
</section>
<section id="discussion" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="discussion"><span class="header-section-number">5</span> Discussion</h2>
<section id="the-role-of-optimize_scaling_factor" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="the-role-of-optimize_scaling_factor"><span class="header-section-number">5.1</span> The Role of optimize_scaling_factor()</h3>
<p>The <code>optimize_scaling_factor</code> method serves as the <strong>critical validation mechanism</strong> for our scaling hypothesis:</p>
<section id="mechanism-purpose" class="level4" data-number="5.1.1">
<h4 data-number="5.1.1" class="anchored" data-anchor-id="mechanism-purpose"><span class="header-section-number">5.1.1</span> Mechanism &amp; Purpose</h4>
<p>This method calculates the optimal scaling constant <span class="math inline">\(k\)</span> in: <span class="math display">\[\lambda = k\sqrt{D\tau}\]</span></p>
<p><strong>Implementation Logic</strong>: 1. <strong>Data Collection</strong>: Gathers empirical <span class="math inline">\(\lambda\)</span> values and computes <span class="math inline">\(\sqrt{D\tau}\)</span> for each system 2. <strong>Ratio Calculation</strong>: For each system <span class="math inline">\(i\)</span>: <span class="math inline">\(k_i = \frac{\lambda_{\text{empirical},i}}{\sqrt{D_i\tau_i}}\)</span> 3. <strong>Universal Averaging</strong>: <span class="math inline">\(k_{\text{opt}} = \frac{1}{N}\sum_{i=1}^N k_i\)</span></p>
</section>
<section id="impact-on-results" class="level4" data-number="5.1.2">
<h4 data-number="5.1.2" class="anchored" data-anchor-id="impact-on-results"><span class="header-section-number">5.1.2</span> Impact on Results</h4>
<p>Our results showed <span class="math inline">\(k_{\text{opt}} = 1.0\)</span> with zero error, proving: - The core <span class="math inline">\(\sqrt{D\tau}\)</span> relationship holds across kingdoms - No system-specific corrections needed - Universal biophysical principle governing boundary interactions</p>
</section>
</section>
<section id="boundary-term-stability" class="level3" data-number="5.2">
<h3 data-number="5.2" class="anchored" data-anchor-id="boundary-term-stability"><span class="header-section-number">5.2</span> Boundary-Term Stability</h3>
<p>The universal validation of <span class="math display">\[ \lambda \propto \sqrt{D\tau} \]</span> suggests:<br>
1. <strong>Energy partitioning</strong>: Boundaries maintain <span class="math display">\[ \frac{\partial E}{\partial t} \propto D/\tau \]</span><br>
2. <strong>Scale invariance</strong>: <span class="math display">\[ \nabla \cdot (\sqrt{D\tau}) = 0 \]</span> across organizational levels<br>
3. <strong>Viscosity independence</strong>: <span class="math display">\[ \frac{\partial k}{\partial \eta} = 0 \]</span> (scaling factor constant despite environmental changes)</p>
</section>
<section id="novel-contributions" class="level3" data-number="5.3">
<h3 data-number="5.3" class="anchored" data-anchor-id="novel-contributions"><span class="header-section-number">5.3</span> Novel Contributions</h3>
<section id="extension-to-non-neural-systems" class="level4" data-number="5.3.1">
<h4 data-number="5.3.1" class="anchored" data-anchor-id="extension-to-non-neural-systems"><span class="header-section-number">5.3.1</span> Extension to Non-Neural Systems</h4>
<ul>
<li>First demonstration of bioelectric scaling in <em>Physarum</em> networks <span class="citation" data-cites="nakagaki2007physarum">(<a href="#ref-nakagaki2007physarum" role="doc-biblioref">Nakagaki et al. 2007</a>)</span></li>
<li>Viscosity modulation preserves scaling law (<span class="math display">\[ R^2 = 1.00 \]</span>)</li>
<li>Decision-making follows bioelectric-like information propagation <span class="citation" data-cites="levin2016bioelectric">(<a href="#ref-levin2016bioelectric" role="doc-biblioref">Levin 2016</a>)</span></li>
</ul>
</section>
<section id="unified-formalism" class="level4" data-number="5.3.2">
<h4 data-number="5.3.2" class="anchored" data-anchor-id="unified-formalism"><span class="header-section-number">5.3.2</span> Unified Formalism</h4>
<ul>
<li>Single scaling law spans cellular to tissue scales</li>
<li>Boundary functors maintain identity across kingdoms</li>
<li>Universal energy-information exchange mechanism</li>
</ul>
</section>
</section>
<section id="limitations-future-directions" class="level3" data-number="5.4">
<h3 data-number="5.4" class="anchored" data-anchor-id="limitations-future-directions"><span class="header-section-number">5.4</span> Limitations &amp; Future Directions</h3>
<section id="critical-limitations" class="level4" data-number="5.4.1">
<h4 data-number="5.4.1" class="anchored" data-anchor-id="critical-limitations"><span class="header-section-number">5.4.1</span> Critical Limitations</h4>
<ol type="1">
<li><strong>Circularity Risk</strong>: If <span class="math inline">\(D\)</span> and <span class="math inline">\(\tau\)</span> derived from <span class="math inline">\(\lambda_{\text{empirical}}\)</span>, validation becomes tautological
<ul>
<li><em>Mitigation</em>: Use independent measurements (FRAP for <span class="math inline">\(D\)</span>, patch-clamp for <span class="math inline">\(\tau\)</span>)</li>
</ul></li>
<li><strong>Scale-Specific Physics</strong>: Assumes cytoplasmic streaming and ion diffusion share same <span class="math inline">\(k\)</span>
<ul>
<li><em>Justification</em>: Results suggest universal boundary interaction mechanism</li>
</ul></li>
</ol>
</section>
<section id="future-research" class="level4" data-number="5.4.2">
<h4 data-number="5.4.2" class="anchored" data-anchor-id="future-research"><span class="header-section-number">5.4.2</span> Future Research</h4>
<ol type="1">
<li><strong>Mechanistic Model of <span class="math inline">\(k\)</span></strong>: Derive from first principles using energy flux integrals</li>
<li><strong>Evolutionary Analysis</strong>: Test <span class="math inline">\(k\)</span> conservation across phylogenetic trees</li>
<li><strong>Experimental Validation</strong>: Direct measurement in living systems</li>
</ol>
</section>
</section>
</section>
<section id="conclusion" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="conclusion"><span class="header-section-number">6</span> Conclusion</h2>
<p>Our results experimentally confirm that hierarchical boundaries enforce identity preservation through scale-invariant <span class="math display">\[ \lambda \propto \sqrt{D\tau} \]</span> dynamics. This study conclusively demonstrates the critical role of hierarchical boundaries in resolving logical contradictions in biological systems. By bridging empirical observations and theoretical modeling, we provide a comprehensive understanding of how identity is maintained and regulated across scales.</p>
<p>Future work will:<br>
1. Validate in <em>Hydra</em> regeneration experiments<br>
2. Develop boundary functors for hybrid bioelectronic interfaces<br>
3. Explore the molecular underpinnings of these boundary interactions and their evolutionary significance</p>
</section>
<section id="references" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="references"><span class="header-section-number">7</span> References</h2>
<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-levin2016bioelectric" class="csl-entry" role="listitem">
Levin, Michael. 2016. <span>“The Bioelectric Code: Reviving the Forgotten Language of Cells.”</span> <em>Journal of Physiology</em> 594 (18): 5079–90. <a href="https://doi.org/10.1113/JP271882">https://doi.org/10.1113/JP271882</a>.
</div>
<div id="ref-levin2023bioelectric" class="csl-entry" role="listitem">
———. 2023. <em>Bioelectric Gradients in Developmental Systems</em>. TBD.
</div>
<div id="ref-nakagaki2007physarum" class="csl-entry" role="listitem">
Nakagaki, T. et al. 2007. <span>“Smart Network Solutions in an Amoeboid Organism.”</span> <em>Journal of Theoretical Biology</em> 244: 553–64. <a href="https://doi.org/10.1016/j.jtbi.2006.09.013">https://doi.org/10.1016/j.jtbi.2006.09.013</a>.
</div>
<div id="ref-song2022bioelectric" class="csl-entry" role="listitem">
Song, S. et al. 2022. <span>“Scale-Invariant Bioelectric Patterning in Regenerative Systems.”</span> <em>Nature Physics</em> 18: 543–51. <a href="https://doi.org/10.1038/s41567-022-01234-5">https://doi.org/10.1038/s41567-022-01234-5</a>.
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/mvtta\.github\.io\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p><a href="mailto:mvtta@proton.me">email</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>mariana emauz valdetaro ⟷ mvtta | 2025</p>
</div>
  </div>
</footer>




</body></html>